File: src/service/UProgress.Service/Services/TaskAnswerService.cs
using Microsoft.EntityFrameworkCore;
using UProgress.Contracts.Models;
using UProgress.Service.Repositories;

namespace UProgress.Service.Services;

public class TaskAnswerService
{
    private readonly TaskAnswerRepository _taskAnswerRepository;
    private readonly AnswerHistoryRepository _answerHistoryRepository;
    private readonly UnitOfWork _unitOfWork;

    public TaskAnswerService(TaskAnswerRepository taskAnswerRepository, UnitOfWork unitOfWork,
        AnswerHistoryRepository answerHistoryRepository)
    {
        _taskAnswerRepository = taskAnswerRepository;
        _unitOfWork = unitOfWork;
        _answerHistoryRepository = answerHistoryRepository;
    }

    public async Task<Guid> CreateTaskAnswer(string answer, Guid taskId, Guid studentId)
    {
        var taskAnswer = new TaskAnswer
        {
            Answer = answer,
            Status = AnswerStatus.Draft,
            TaskId = taskId,
            StudentId = studentId
        };

        _taskAnswerRepository.Insert(taskAnswer);
        await _unitOfWork.SaveAsync();

        return taskAnswer.Id;
    }

    public async Task<Guid> CreateAnswerHistory(AnswerStatus statusMovedTo, Guid userId, Guid answerId,
        string? message = null)
    {
        var answerHistory = new AnswerHistory
        {
            Date = DateTime.UtcNow,
            StatusMovedTo = statusMovedTo,
            Message = message,
            UserId = userId,
            AnswerId = answerId
        };

        _answerHistoryRepository.Insert(answerHistory);
        await _unitOfWork.SaveAsync();

        return answerHistory.Id;
    }

    public async Task<bool> RequestApprove(Guid taskAnswerId, string? message, string answer)
    {
        var taskAnswer = _taskAnswerRepository.Get().Include(ta => ta.History)
            .FirstOrDefault(ta => ta.Id == taskAnswerId);
        if (taskAnswer == null)
        {
            return false;
        }

        taskAnswer.Status = AnswerStatus.Approvable;
        taskAnswer.Answer = answer;
        taskAnswer.History.Add(new AnswerHistory
        {
            Date = DateTime.UtcNow,
            StatusMovedTo = AnswerStatus.Approvable,
            Message = message,
            UserId = taskAnswer.StudentId,
            AnswerId = taskAnswer.Id
        });

        _taskAnswerRepository.Update(taskAnswer);
        await _unitOfWork.SaveAsync();

        return true;
    }

    public async Task<bool> Approve(Guid taskAnswerId, Guid teacherId, string? message, int mark)
    {
        var taskAnswer = _taskAnswerRepository.Get().Include(ta => ta.History)
            .FirstOrDefault(ta => ta.Id == taskAnswerId);
        if (taskAnswer == null)
        {
            return false;
        }

        taskAnswer.Status = AnswerStatus.Approved;
        taskAnswer.Mark = mark;
        taskAnswer.ApprovedById = teacherId;
        taskAnswer.History.Add(new AnswerHistory
        {
            Date = DateTime.UtcNow,
            StatusMovedTo = AnswerStatus.Approved,
            Message = message,
            UserId = teacherId,
            AnswerId = taskAnswer.Id
        });

        _taskAnswerRepository.Update(taskAnswer);
        await _unitOfWork.SaveAsync();

        return true;
    }

    public async Task<bool> Reject(Guid taskAnswerId, Guid teacherId, string? message)
    {
        var taskAnswer = _taskAnswerRepository.Get().Include(ta => ta.History)
            .FirstOrDefault(ta => ta.Id == taskAnswerId);
        if (taskAnswer == null)
        {
            return false;
        }

        taskAnswer.Status = AnswerStatus.Rejected;
        taskAnswer.History.Add(new AnswerHistory
        {
            Date = DateTime.UtcNow,
            StatusMovedTo = AnswerStatus.Rejected,
            Message = message,
            UserId = teacherId,
            AnswerId = taskAnswer.Id
        });

        _taskAnswerRepository.Update(taskAnswer);
        await _unitOfWork.SaveAsync();

        return true;
    }
}

File: src/ui/packages/uprogress/src/services/taskAnswerService.ts
import {
  GetTaskAnswer,
  GetTaskAnswerResult,
} from './../models/messages/GetTaskAnswer';
import { API_URLS } from '../constants';
import { getHttpClient } from '../utils/httpUtils';
import {
  CreateTaskAnswer,
  CreateTaskAnswerResult,
} from './../models/messages/CreateTaskAnswer';
import {
  RequestApprove,
  RequestApproveResult,
} from '../models/messages/RequestApprove';
import { Approve, ApproveResult } from '../models/messages/Approve';
import { Reject, RejectResult } from '../models/messages/Reject';

const createTaskAnswer = async (
  message: CreateTaskAnswer
): Promise<CreateTaskAnswerResult> => {
  try {
    const result = (
      await getHttpClient().post(API_URLS.taskAnswer.createTaskAnswer, message)
    ).data;

    return {
      isSuccessful: true,
      taskAnswerId: result.taskAnswerId,
    };
  } catch (e: unknown) {
    return {
      isSuccessful: false,
    } as CreateTaskAnswerResult;
  }
};

const getTaskAnswer = async (
  message: GetTaskAnswer
): Promise<GetTaskAnswerResult> => {
  try {
    const result = (
      await getHttpClient().post(API_URLS.taskAnswer.getTaskAnswer, message)
    ).data;

    return {
      isSuccessful: true,
      id: result.id,
      mark: result.mark,
      answer: result.answer,
      status: result.status,
      taskId: result.taskId,
      taskName: result.taskName,
      studentId: result.studentId,
      approvedById: result.approvedById,
      disciplineId: result.disciplineId,
      disciplineName: result.disciplineName,
      history: result.history,
      attachments: result.attachments,
    };
  } catch (e: unknown) {
    return {
      isSuccessful: false,
    } as GetTaskAnswerResult;
  }
};

const requestApprove = async (
  message: RequestApprove
): Promise<RequestApproveResult> => {
  try {
    await getHttpClient().post(API_URLS.taskAnswer.requestApprove, message);

    return {
      isSuccessful: true,
    };
  } catch (e: unknown) {
    return {
      isSuccessful: false,
    } as RequestApproveResult;
  }
};

const approve = async (message: Approve): Promise<ApproveResult> => {
  try {
    await getHttpClient().post(API_URLS.taskAnswer.approve, message);

    return {
      isSuccessful: true,
    };
  } catch (e: unknown) {
    return {
      isSuccessful: false,
    } as ApproveResult;
  }
};

const reject = async (message: Reject): Promise<RejectResult> => {
  try {
    await getHttpClient().post(API_URLS.taskAnswer.reject, message);

    return {
      isSuccessful: true,
    };
  } catch (e: unknown) {
    return {
      isSuccessful: false,
    } as RejectResult;
  }
};

export const taskAnswerService = {
  createTaskAnswer,
  getTaskAnswer,
  requestApprove,
  approve,
  reject,
};
